Enhancements, optimization for ex37

thatguynoe opened the pull request:
* Move shared Bregman projection function into header file.
* Add choice to control growth rate for gradient descent step length.
* Move some assembly outside of the solve iteration loop.
* ~~Add choice to use the bisection method for root-finding instead of Newton's method.~~
* Use the [Illinois method](https://en.wikipedia.org/wiki/Regula_falsi#The_Illinois_algorithm) for root-finding instead of Newton's method.

<!--GHEX{"author":"thatguynoe","assignment":"2025-07-01T21:27:56","editor":"tzanio","id":4645,"reviewers":["vladotomov","bslazarov"],"merge":"2026-04-19T00:55:19","approval":"2026-04-14T19:39:23"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4645](https://github.com/mfem/mfem/pull/4645) | @thatguynoe | @tzanio | @vladotomov + @bslazarov | 7/1/25 | 4/14/26 | 4/18/26 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

thatguynoe committed:
move proj function into header file

thatguynoe committed:
add growth rate arg for grad descent step length

thatguynoe committed:
assemble boundary, bilinear form outside of solve

thatguynoe committed:
assemble boundary, bilinear form outside of solve

thatguynoe committed:
add option to choose bisection method for roots

thatguynoe committed:
set smaller itol for better convergence

thatguynoe committed:
move ParGridFunction declaration

thatguynoe committed:
fix ex37 serial

If running ex37 serial with MFEM parallel, a segfault would occur when attempting to run MPI_Allreduce. To fix this, we use the associated FiniteElementSpace and check for MFEM parallel.

thatguynoe committed:
fix ex37 serial

If running ex37 serial with MFEM parallel, a segfault would occur when attempting to run MPI_Allreduce. To fix this, we use the associated FiniteElementSpace and check for MFEM parallel.

thatguynoe committed:
delete bilinear form when cleaning

brendankeith commented on examples/ex37.cpp:
Please update the references here and in `ex37p.cpp`

brendankeith replied:
Example code `ex37 -lambda 0.1 -mu 0.1` is no longer stable

brendankeith replied:
`ex37 -o 2 -alpha 5.0 -mi 50 -vf 0.4 -ntol 1e-5` shows strange artefacts with the new default growth rate.

I suggest
```ex37 -o 2 -alpha 10.0 -mi 50 -vf 0.4 -ntol 1e-5 -growth 1.5```

brendankeith commented on examples/ex37.cpp:
```diff
@@ -180,16 +133,18 @@ int main(int argc, char *argv[])
    real_t mu = 1.0;
    bool glvis_visualization = true;
    bool paraview_output = false;
+   bool newton = true;
```
Suggestion: Use a string such as `root_finding_alg` that can be either `newton` or `bisection`. It is not clear until reading on what `newton` is used for

brendankeith replied:
See also https://github.com/mfem/mfem/pull/4645#discussion_r1898110950

thatguynoe replied:
In this case, it is necessary to check for valid input so that either `root_finding_alg == "newton"` or `root_finding_alg == "bisection"`. Would it be make more sense to implement it as part of the [args check](https://github.com/mfem/mfem/blob/d7681c26dc608bfab1ca4c891e22cbc1260340c6/examples/ex37p.cpp#L188), or is it sufficient to change [this `else if`](https://github.com/mfem/mfem/blob/d7681c26dc608bfab1ca4c891e22cbc1260340c6/examples/ex37p.cpp#L415) to an `else`?

brendankeith commented on examples/ex37p.cpp:
Example code `mpirun -np 4 ex37p -lambda 0.1 -mu 0.1` is unstable with the new growth rate

brendankeith replied:
Update `mpirun -np 4 ex37p -o 2 -alpha 5.0 -mi 50 -vf 0.4 -ntol 1e-5` as in `ex37.cpp`

brendankeith commented on examples/ex37.hpp:
General topics for discussion:

1. Do we need two different root-finding algorithms in this example? I think it is better to just have one
2. We now know that searching for the root of `f(µ) = ∫_Ω sigmoid(ψ - αµ) dx - θ vol(Ω)` is more stable than the root of `f(c) = ∫_Ω sigmoid(ψ + c) dx - θ vol(Ω)`. Why don't we do this instead?

thatguynoe replied:
Changes implemented in https://github.com/mfem/mfem/pull/4645/commits/5808fc6966faf0d9b1a34c04f1654eae5bcfa2b4.

thatguynoe committed:
update sample runs, fix stability

thatguynoe committed:
rename newton arg for clarity

thatguynoe committed:
Revert "rename newton arg for clarity"

This reverts commit d7681c26dc608bfab1ca4c891e22cbc1260340c6.

thatguynoe committed:
Revert "add option to choose bisection method for roots"

This reverts commit 21778cd9335337a1419af482a7feaf6baac1057f.

thatguynoe committed:
use gradient descent step length in proj

See https://github.com/mfem/mfem/pull/4645#discussion_r1898110950.

brendankeith committed:
Merge branch 'master' into ex37

brendankeith commented on examples/ex37.hpp:
```diff
@@ -422,12 +513,8 @@ void DiffusionSolver::SetupFEM()
 }
 
-void DiffusionSolver::Solve()
+void DiffusionSolver::AssembleBoundary()
```
Why is this a separate function now?

brendankeith replied:
I suggest combining the BC function into the bilinear form assembly function with `update_BCs` as an optional boolean argument. This is safer for the end user who could accidentally call these functions in the wrong order.

brendankeith commented:
@thatguynoe, I put the `WIP` label back because I noticed one of the items supposed to be in this PR is using the bisection method instead of Newton for root-finding

thatguynoe committed:
Merge branch 'master' into ex37

thatguynoe committed:
use the bisection method instead of Newton

thatguynoe commented:
> @thatguynoe, I put the `WIP` label back because I noticed one of the items supposed to be in this PR is using the bisection method instead of Newton for root-finding

I misunderstood which we wanted to use -- addressed in https://github.com/mfem/mfem/pull/4645/commits/2a4decc63536247c4b5948171b07419936696f72.

thatguynoe committed:
move boundary assembly into bilinear form assembly

Prevents the user from accidentally calling these methods in the wrong order.

wasness-apma reviewed (commented):
For the Bregman projection, you can use the Illinois method rather than bisection to speed up convergence. This would look more like:

```cpp
{
c = b - f_b * (b - a) / (f_b - f_a); // f_a, f_b saved from before, and have already had target_volume subtracted
y = c;
f_c = int_sigmoid_psi->Sum();
f_c -= target_volume;

if ((f_c == 0) || abs(b - a) / 2.0 < tol) { done = true; y = 0; break; }

if (sign(f_a) == sign(f_c)) {
    a = b;
    f_a = f_b;
} else {
    f_a = f_a / 2;
};
b = c;
f_b = f_c;
// loop
};
```

Also, as a side note, the function as implemented has a redundant computation for f_a that could be done more cleanly by saving the previous computations, leading to a single domain integration in each loop rather than two. Furthermore, we can run the (a - b) / 2.0 < tol test before doing any domain integrations to save an iteration.

tzanio committed:
Merge branch 'master' into ex37

thatguynoe committed:
use the Illinois method instead of bisection

Speeds up convergence for the Bregman projection.

thatguynoe committed:
remove signum function

thatguynoe committed:
Merge branch 'master' into ex37

tzanio committed:
Merge branch 'master' into ex37

vladotomov reviewed (approved):
ping @bslazarov

dohyun-cse commented on examples/ex37.hpp:
```diff
@@ -371,6 +375,130 @@ class LinearElasticitySolver
+#ifdef MFEM_USE_MPI
+   if (pfes)
+   {
+      MPI_Allreduce(MPI_IN_PLACE, &f_a, 1, MFEM_MPI_REAL_T,
```
It seems that most examples use `MPITypeMap<real_t>::mpi_type` instead of `MFEM_MPI_REAL_T`. For consistency, please consider the following change (and other`MPI` calls)
```suggestion
      MPI_Allreduce(MPI_IN_PLACE, &f_a, 1, MPITypeMap<real_t>::mpi_type,
```

thatguynoe replied:
Addressed in https://github.com/mfem/mfem/pull/4645/commits/b863dd186f178f6f4644e38d921be0aa46bb877e.

dohyun-cse commented on examples/ex37.hpp:
```diff
@@ -422,12 +550,8 @@ void DiffusionSolver::SetupFEM()
 }
 
-void DiffusionSolver::Solve()
+void DiffusionSolver::AssembleBoundary()
```
The name is misleading. It sounds like this function assembles boundary bilinear form integrators, not extracting boundary dofs. Please consider changing it to `UpdateEssentialTDofs` or something else.

dohyun-cse commented on examples/ex37.hpp:
```diff
@@ -440,7 +564,39 @@ void DiffusionSolver::Solve()
+   OperatorPtr A;
+   Vector B, X;
+
    if (b)
```
Do we need to delete `b`?  When we call `LinearForm::Assemble`, then the data will be reset to 0 and assemble the integrators by default. This is only necessary when we change the integrators related to this linear form. So, reallocating `b` is not necessary.

thatguynoe committed:
correct return type in proj function

Co-authored-by: Dohyun Kim <dhkim.cse@gmail.com>

thatguynoe committed:
use MPITypeMap<real_t>::mpi_type

thatguynoe committed:
rename boundary dof extraction function

thatguynoe committed:
rename bool flag

thatguynoe committed:
Merge branch 'master' into ex37

thatguynoe committed:
store the diffusion global matrix

bslazarov commented on examples/ex37.hpp:
```diff
@@ -371,6 +376,130 @@ class LinearElasticitySolver
+      mfem_warning("Projection reached maximum iteration without converging. "
+                   "Result may not be accurate.");
+   }
+   psi += c;
```
Shouldn't y be set to y=0.0;

thatguynoe committed:
set y = 0.0

thatguynoe committed:
correct comments

thatguynoe committed:
ensure root search interval is valid

We now look for a root of the function f(c) = ∫_Ω sigmoid(ψ + c) dx - θ vol(Ω) within the interval [a,b], where a := -‖αG‖_∞ and b := ‖αG‖_∞, and α and G are as in Step 4 and Step 5. It follows that f(a) ≤ ∫_Ω sigmoid(ψ + αG) dx - θ vol(Ω) ≤ f(b), and the inner quantity equals 0 since ψ_new := ψ_prev - αG in Step 5 and ∫_Ω sigmoid(ψ_new) dx - θ vol(Ω) = 0. This ensures f(a) ≤ 0 ≤ f(b), as required by the Illinois method.

thatguynoe committed:
correct assert message

thatguynoe committed:
parameter c no longer present

thatguynoe committed:
Merge branch 'master' into ex37

thatguynoe committed:
initialize y, c, f_c

tzanio commented:
Merged in `next` for testing...
